File: python-app/models/pivot_table/self.py
import pandas as pd

from lib.pivot_table.plot_data import plot_data

from models.pivot_table.pivot_table_level import PivotTableLevel, level_to_spanish
from models.pivot_table.aggregate_function_type import AggregateFunctionType, aggregate_function_to_spanish
from models.pivot_table.filter_function_type import FilterFunctionType, filter_function_to_spanish
from models.pivot_table.aggregate_function_type import AggregateFunctionType
from models.pivot_table.filter_function_type import FilterFunctionType
from models.pivot_table.pivot_table_level import PivotTableLevel
from models.pivot_table.data_source import DataSource
from models.pivot_table.data_filter.self import DataFilter
from models.pivot_table.data_filter.charting_mode import ChartingMode
from models.slide.slide_category import SlideCategory
from models.slide.self import Slide

from render.grid_layout.self import grid_layout
from render.grid_layout.grid_layout_item import GridLayoutItem

from render.fractional_unit import FractionalUnit
from render.widget.text_widget import TextWidget
from render.widget.image_widget import ImageWidget
from render.font import Font
from render.color import Color

from pptx.util import Cm, Pt
from pptx.enum.text import PP_ALIGN

class PivotTable(Slide):
    def __init__(
            self, 
            title: str, 
            identifier: str, 
            creation_date: pd.Timestamp, 
            last_edit: pd.Timestamp,
            preview: str,
            bare_preview: str,
            filters: list[DataFilter],
            filters_order: list[PivotTableLevel],
            source: DataSource,
            data: dict[str, dict[str, float]] | dict[str, float],
            aggregate_function: AggregateFunctionType,
            filter_function: FilterFunctionType,
            ) -> None:
        super().__init__(
            title=title, 
            identifier=identifier, 
            creation_date=creation_date, 
            last_edit=last_edit, 
            preview=preview, 
            category=SlideCategory.PIVOT_TABLE
            )

        self.bare_preview = bare_preview
        self.filters: list[DataFilter] = filters
        self.filters_order = filters_order
        self.source = source
        self.data = data
        self.aggregate_function = aggregate_function
        self.filter_function = filter_function

    def to_dict(self) -> dict:
        return {
            **super().to_dict(),
            "bare_preview": self.bare_preview,
            "filters": [f.to_dict() for f in self.filters],
            "filters_order": [ level.value for level in self.filters_order ],
            "source": self.source.to_dict(),
            "data": self.data,
            "aggregate_function": self.aggregate_function.value,
            "filter_function": self.filter_function.value,
        }
    
    def render_bare_preview(self, filepath: str):
        outer_filter: DataFilter = next((_filter for _filter in self.filters if _filter.charting_mode == ChartingMode.SUPER_CHART), None)
        if outer_filter is None:
            outer_filter = next((_filter for _filter in self.filters if _filter.charting_mode == ChartingMode.CHART), None)

        plot_data(
            data=self.data, 
            title=self.title, 
            kind="bar", 
            x_label=level_to_spanish(outer_filter.level), 
            y_label=
                aggregate_function_to_spanish(self.aggregate_function) + 
                " de calificaciones de " + 
                filter_function_to_spanish(outer_filter),
            filepath=filepath
            )
        
        self.bare_preview = filepath
    
    def render(self, slide, drawable_area):
        grid_layout(
            slide=slide,
            drawable_area=drawable_area,
            grid_areas='''
            a
            b
            ''',
            column_widths=[FractionalUnit(1)],
            row_heights=[
                Cm(3.5).emu, 
                FractionalUnit(1) 
            ],
            gap=Cm(0.25).emu,
            children=[
                GridLayoutItem(
                    area='a',
                    child=TextWidget(
                        text=self.title, 
                        alignment=PP_ALIGN.CENTER,
                        font=Font(
                            size=Pt(60),
                            bold=True,
                            font_family='Avenir'
                        ),
                        color=Color(0x1F, 0x38, 0x64)
                    )
                ),
                GridLayoutItem(
                    area='b',
                    child=ImageWidget(source=self.bare_preview)
                ),
            ]
        )

File: lib/models/pivot_table/self.dart
import 'package:ut_report_generator/models/pivot_table/aggregate_function_type.dart';
import 'package:ut_report_generator/models/pivot_table/data_filter/self.dart';
import 'package:ut_report_generator/models/pivot_table/data_source.dart';
import 'package:ut_report_generator/models/pivot_table/filter_function_type.dart';
import 'package:ut_report_generator/models/pivot_table/pivot_table_data.dart';
import 'package:ut_report_generator/models/pivot_table/pivot_table_level.dart';
import 'package:ut_report_generator/models/slide/self.dart';
import 'package:ut_report_generator/models/slide_category.dart';

class PivotTable extends Slide {
  final String barePreview;
  final DataSource source;
  final List<DataFilter> filters;
  final List<PivotTableLevel> filtersOrder;
  final PivotTableData data;
  final AggregateFunctionType aggregateFunction;
  final FilterFunctionType filterFunction;
  final bool automaticTitle;

  PivotTable({
    required super.title,
    required super.identifier,
    required super.creationDate,
    required super.lastEdit,
    required super.preview,
    required this.barePreview,
    required this.source,
    required this.filters,
    required this.filtersOrder,
    required this.data,
    required this.aggregateFunction,
    required this.filterFunction,
    required this.automaticTitle,
  }) : super(category: SlideCategory.pivotTable);

  Map<String, dynamic> toJson() {
    return {
      ...super.toMap(),
      "bare_preview": barePreview,
      "source": source.toJson(),
      "filters": filters.map((f) => f.toJson()).toList(),
      "filters_order": filtersOrder.map((value) => value.name).toList(),
      "data": data,
      "aggregate_function": aggregateFunction.name,
      "filter_function": filterFunction.name,
    };
  }

  factory PivotTable.fromJson(Map<String, dynamic> json) {
    final title = json["title"];
    final identifier = json["identifier"];
    final creationDate = DateTime.parse(json["creation_date"]);
    final lastEdit = DateTime.parse(json["last_edit"]);
    final preview = json["preview"];
    final barePreview = json["bare_preview"];
    final source = DataSource.fromJson(json["source"]);
    final filters =
        (json["filters"] as List).map((e) => DataFilter.fromJson(e)).toList();
    final filtersOrder =
        (json["filters_order"] as List)
            .map((e) => PivotTableLevel.values.byName(e))
            .toList();
    final data = PivotTableData.fromJson(json["data"]);
    final aggregateFunction = AggregateFunctionType.values.byName(
      json["aggregate_function"],
    );
    final filterFunction = FilterFunctionType.values.byName(
      json["filter_function"],
    );

    return PivotTable(
      title: title,
      identifier: identifier,
      creationDate: creationDate,
      lastEdit: lastEdit,
      preview: preview,
      barePreview: barePreview,
      source: source,
      filters: filters,
      filtersOrder: filtersOrder,
      data: data,
      aggregateFunction: aggregateFunction,
      filterFunction: filterFunction,
      automaticTitle: true,
    );
  }

  PivotTable copyWith({
    String? title,
    String? identifier,
    DateTime? creationDate,
    DateTime? lastEdit,
    String? barePreview,
    String? preview,
    DataSource? source,
    List<DataFilter>? filters,
    List<PivotTableLevel>? filtersOrder,
    PivotTableData? data,
    AggregateFunctionType? aggregateFunction,
    FilterFunctionType? filterFunction,
    bool? automaticTitle,
  }) {
    return PivotTable(
      title: title ?? this.title,
      identifier: identifier ?? this.identifier,
      creationDate: creationDate ?? this.creationDate,
      lastEdit: lastEdit ?? this.lastEdit,
      preview: preview ?? this.preview,
      barePreview: barePreview ?? this.barePreview,
      source: source ?? this.source,
      filters: filters ?? this.filters,
      filtersOrder: filtersOrder ?? this.filtersOrder,
      data: data ?? this.data,
      aggregateFunction: aggregateFunction ?? this.aggregateFunction,
      filterFunction: filterFunction ?? this.filterFunction,
      automaticTitle: automaticTitle ?? this.automaticTitle,
    );
  }
}
